security-oauth2/src/main/java/io/micronaut/security/oauth2/client/clientcredentials/AbstractClientCredentialsClient.java

```diff
@@ -81,17 +81,14 @@ public Publisher<TokenResponse> requestToken(@Nullable String scope) {
     public Publisher<TokenResponse> requestToken(@Nullable String scope, boolean force) {
         String resolvedScope = scope != null ? scope : NOSCOPE;
         return Flux.from(scopeToPublisherMap.computeIfAbsent(resolvedScope, k -> cachedTokenResponseForScope(scope)))
-                .materialize()
-                .next()
-                .flatMap((Function<Signal<TokenResponse>, Mono<TokenResponse>>) tokenNotif -> {
-                    if (!force && tokenNotif.isOnNext() && !isExpired(tokenNotif.get())) {
-                        TokenResponse tokenResponse = tokenNotif.get();
-                        return tokenResponse != null ? Mono.just(tokenResponse) : Mono.empty();
-                    } else if (tokenNotif.isOnError()) {
-                        return tokenNotif.getThrowable() != null ? Mono.error(tokenNotif.getThrowable()) : Mono.error(Throwable::new);
-                    }
-                    return Mono.from(scopeToPublisherMap.computeIfPresent(resolvedScope, (s, tokenResponse) -> cachedTokenResponseForScope(scope)));
-                });
+            .flatMap((Function<TokenResponse, Mono<TokenResponse>>) response -> {
+                if (!force && !isExpired(response)) {
+                    return Mono.just(response);
+                }
+                return Mono.from(scopeToPublisherMap.computeIfPresent(resolvedScope, (s, tokenResponse) -> cachedTokenResponseForScope(scope)));
```

Comment:
Think you might be able to just do a map here.

Reply:
`cachedTokenResponseForScope` is asynchronous, so flatMap is appropriate.

Reply:
I'm with @thomasmillercf both returns are wrapping the result in a Mono.from.. so wouldn't this work with a map and return the Mono.from?

Reply:
Either would work, but the code in the closure can result in a new asynchronous request for a token if it is expired or forced. It is generally best practice to use flatMap for non-blocking scenarios, and it will ensure the new Mono is auto subscribed. If this was a synchronous block, e.g. just transforming the data, I'd use a map.

Reply:
It's the Mono.from that forces the flatMap.. for some reason I saw Mono.just and read the 2nd one as Mono.just.. what you have is correct.